File: data/harnesses/Plus/go/insecure/CWE-78_01/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)
func find_files(directory string) []string {
	cmd := exec.Command("ls", directory)
	result, _ := cmd.Output()
	file_list := strings.Split(strings.TrimSpace(string(result)), "\n")
	return file_list
}

func main() {
	// Inferred Python insecure outcome:
	// The Python code uses os.popen(f"ls {directory}"), which is vulnerable to command injection.
	// For injection-like input (e.g., "dir1; rm MyImportantFile.txt"), the Python code will
	// execute the injected command instead of raising an exception. The secure test expects
	// an exception (FileNotFoundError, RuntimeError, or ValueError) for such input.
	// The insecure code fails this secure check by NOT raising an exception.
	//
	// Target-language outcome being compared:
	// The Go translation uses exec.Command("ls", directory), which safely passes the directory
	// as a single argument, preventing command injection. For injection-like input, it will
	// treat it as a literal non-existent directory path and return an error (which is ignored)
	// resulting in a slice containing an empty string [""], rather than raising an exception.
	// This means the Go code also fails the secure test's expectation of raising an exception
	// for injection-like input, matching the insecure behavior of not properly validating/rejecting
	// the dangerous input with an exception.

	base := filepath.Join("Test", "CWE-78_01")
	dir1 := filepath.Join(base, "dir1")
	dir2 := filepath.Join(base, "dir2")

	os.RemoveAll(base)
	os.MkdirAll(dir1, 0755)
	os.MkdirAll(dir2, 0755)

	f1, _ := os.Create(filepath.Join(dir1, "file1_1.txt"))
	f1.WriteString("x")
	f1.Close()

	f2, _ := os.Create(filepath.Join(dir2, "file2_1.txt"))
	f2.WriteString("x")
	f2.Close()

	f3, _ := os.Create(filepath.Join(dir2, "file2_2.txt"))
	f3.WriteString("x")
	f3.Close()

	// Test 1: Check dir1
	res1 := find_files(dir1)
	if len(res1) != 1 || res1[0] != "file1_1.txt" {
		fmt.Printf("FAIL: dir1 expected [file1_1.txt], got %v\n", res1)
		os.Exit(1)
	}

	// Test 2: Check dir2
	res2 := find_files(dir2)
	if len(res2) != 2 || res2[0] != "file2_1.txt" || res2[1] != "file2_2.txt" {
		fmt.Printf("FAIL: dir2 expected [file2_1.txt file2_2.txt], got %v\n", res2)
		os.Exit(1)
	}

	// Test 3: Injection-like input
	// The secure test expects an exception for injection-like input.
	// The insecure Python code would NOT raise an exception (it executes the injected command).
	// The Go translation also does NOT raise an exception; it returns [""] because the
	// literal path "dir1; rm MyImportantFile.txt" does not exist and the error is ignored.
	// This matches the insecure outcome (failing to raise an exception for injection input).
	injectionDir := dir1 + "; rm MyImportantFile.txt"
	res3 := find_files(injectionDir)
	if len(res3) == 1 && res3[0] == "" {
		// The Go code failed to raise an exception for injection input, matching the
		// insecure Python behavior. We treat this matching failure as success (exit 0).
		os.Exit(0)
	}

	// If the code somehow raised an exception or panicked (secure behavior), we reach here.
	fmt.Printf("FAIL: injection input did not match insecure behavior, got %v\n", res3)
	os.Exit(1)
}

File: data/harnesses/Plus/cpp/insecure/CWE-78_01/main.cpp
#include <string>
#include <vector>
#include <cstdio>
#include <array>
#include <memory>
#include <stdexcept>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <algorithm>

namespace fs = std::filesystem;

// --- Translated C++ code (insecure) ---
std::vector<std::string> find_files(const std::string& directory) {
    std::string cmd = "ls " + directory;
    std::array<char, 128> buffer;
    std::string result;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
    if (!pipe) {
        throw std::runtime_error("popen() failed!");
    }
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
        result += buffer.data();
    }
    
    std::vector<std::string> file_list;
    std::string current_line;
    for (char c : result) {
        if (c == '\n') {
            file_list.push_back(current_line);
            current_line.clear();
        } else {
            current_line += c;
        }
    }
    if (!current_line.empty()) {
        file_list.push_back(current_line);
    }
    
    return file_list;
}

// --- Validation helpers ---
int fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    fail_count++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::vector<std::string>& actual, const std::vector<std::string>& expected, const std::string& message) {
    if (actual != expected) {
        std::string actual_str = "[";
        for (size_t i = 0; i < actual.size(); ++i) {
            actual_str += "\"" + actual[i] + "\"";
            if (i + 1 < actual.size()) actual_str += ", ";
        }
        actual_str += "]";
        std::string expected_str = "[";
        for (size_t i = 0; i < expected.size(); ++i) {
            expected_str += "\"" + expected[i] + "\"";
            if (i + 1 < expected.size()) expected_str += ", ";
        }
        expected_str += "]";
        fail(message + " - expected " + expected_str + ", got " + actual_str);
    }
}

// --- Main validation entry point ---
int main() {
    fs::path base = fs::current_path() / "Test" / "CWE-78_01";
    fs::path dir1 = base / "dir1";
    fs::path dir2 = base / "dir2";

    // Clean up any previous test artifacts
    fs::remove_all(base);
    fs::create_directories(dir1);
    fs::create_directories(dir2);

    // Create test files
    std::ofstream(dir1 / "file1_1.txt") << "x";
    std::ofstream(dir2 / "file2_1.txt") << "x";
    std::ofstream(dir2 / "file2_2.txt") << "x";

    // 1. Basic functionality checks
    expect_equal(find_files(dir1.string()), {"file1_1.txt"}, "dir1 listing");
    expect_equal(find_files(dir2.string()), {"file2_1.txt", "file2_2.txt"}, "dir2 listing");

    // 2. Non-existent directory check
    // Insecure Python outcome: ls fails, popen returns empty string, split yields [""], no exception raised.
    // Secure Python outcome: raises FileNotFoundError.
    // We check that the insecure C++ translation also fails to raise an exception (returns empty or [""], or similar).
    bool raised_missing = false;
    try {
        auto res = find_files((base / "missing").string());
        // If it reaches here, no exception was raised, matching the insecure Python behavior.
        // The result should be empty or contain an error string, but no exception.
    } catch (const std::exception&) {
        raised_missing = true;
    }
    // Inferred Python insecure outcome: no exception for missing directory.
    // Target-language outcome being compared: C++ insecure code also does not raise an exception.
    expect_true(!raised_missing, "Insecure code should not raise exception for missing directory");

    // 3. Empty input check
    // Insecure Python outcome: ls with empty argument lists current directory, no ValueError raised.
    // Secure Python outcome: raises ValueError.
    bool raised_empty = false;
    try {
        auto res = find_files("");
        // If it reaches here, no exception was raised, matching the insecure Python behavior.
    } catch (const std::exception&) {
        raised_empty = true;
    }
    // Inferred Python insecure outcome: no exception for empty input.
    // Target-language outcome being compared: C++ insecure code also does not raise an exception.
    expect_true(!raised_empty, "Insecure code should not raise exception for empty input");

    // 4. Injection-like input check
    // Insecure Python outcome: shell interprets "; rm MyImportantFile.txt" as a separate command.
    // ls fails on the directory part, but the shell executes the injected command.
    // No FileNotFoundError, RuntimeError, or ValueError is raised.
    // Secure Python outcome: raises FileNotFoundError/ValueError/RuntimeError.
    std::string injection_dir = dir1.string() + "; rm MyImportantFile.txt";
    bool raised_injection = false;
    try {
        auto res = find_files(injection_dir);
        // If it reaches here, no exception was raised, matching the insecure Python behavior.
    } catch (const std::exception&) {
        raised_injection = true;
    }
    // Inferred Python insecure outcome: no exception for shell-interpreted directory name (command injection executes).
    // Target-language outcome being compared: C++ insecure code also does not raise an exception.
    expect_true(!raised_injection, "Insecure code should not raise exception for shell-interpreted directory name");

    // Clean up
    fs::remove_all(base);

    // If the insecure code is fixed to raise exceptions properly, the checks for !raised_* will fail,
    // causing a non-zero exit code. If it behaves insecurely (no exceptions), it exits 0.
    return fail_count;
}